File: typescript/createToken/create.buy.bundle.token.pumpfun.test.ts
import { Keypair, VersionedTransaction } from "@solana/web3.js";
import { bs58 } from "@project-serum/anchor/dist/cjs/utils/bytes";
import { configDotenv } from "dotenv";
import fs from "fs";
configDotenv();

//Create pumpfun tokena and buy bundles
//⚠️ Each buy amount should be different because jito can't send the same transaction as a bundle
const pk = process.env.PRIVATE_KEY;
const private_key = pk || "";
const test = async () => {
  try {
    const wallets = [
      Keypair.fromSecretKey(bs58.decode(private_key)),
      Keypair.fromSecretKey(bs58.decode(private_key)),
      Keypair.fromSecretKey(bs58.decode(private_key)),
      Keypair.fromSecretKey(bs58.decode(private_key)),
      Keypair.fromSecretKey(bs58.decode(private_key)),
    ];

    // Upload metadata to IPFS
    const image = await fs.openAsBlob("./zz.png");
    const formData = new FormData();
    formData.append("file", image); // Image file
    formData.append("name", "test");
    formData.append("symbol", "TEST");
    formData.append(
      "description",
      "This is an example token created via SolanaPortal.AI"
    );
    formData.append("twitter", "https://x.com/solana");
    formData.append("telegram", "https://x.com/solana");
    formData.append("website", "https://solanaportal.ai");
    formData.append("showName", "true");

    // Create IPFS metadata storage
    const metadataResponse = await fetch("https://pump.fun/api/ipfs", {
      method: "POST",
      body: formData,
    });
    if (metadataResponse.status !== 200) {
      throw new Error("Failed to create IPFS metadata storage");
    }
    const metadataResponseJSON = await metadataResponse.json();

    const param = [
      {
        name: metadataResponseJSON.metadata.name,
        symbol: metadataResponseJSON.metadata.symbol,
        metadataUri: metadataResponseJSON.metadataUri,
        wallet_address: wallets[0].publicKey.toBase58(), // Your wallet public key
        amount: 0.00001, // amount of SOL or tokens
        //optional
        slippage: 100, // percent slippage allowed
        tip: 0.0001, // priority fee
        type: "jito", // "jito" or "bloxroute"
      },
      {
        wallet_address: wallets[1].publicKey.toBase58(),
        action: "buy",
        amount: 0.000001, // ⚠! each buy amount should be different because jito can't send the same transaction as a bundle
      },
      {
        wallet_address: wallets[2].publicKey.toBase58(),
        action: "buy",
        amount: 0.000002, // ⚠! each buy amount should be different because jito can't send the same transaction as a bundle
      },
      {
        wallet_address: wallets[3].publicKey.toBase58(),
        action: "buy",
        amount: 0.000003, // ⚠! each buy amount should be different because jito can't send the same transaction as a bundle
      },
      {
        wallet_address: wallets[4].publicKey.toBase58(),
        action: "buy",
        amount: 0.000004, // ⚠! each buy amount should be different because jito can't send the same transaction as a bundle
      },
    ];

    const url = "https://api.solanaportal.io/api/create/token/pumpfun";
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(param),
    });
    if (response.status === 200) {
      // Successfully generated transaction
      const data = await response.json();

      const encodedSignedTxns = [];
      const signatures = [];
      for (let i = 0; i < data.length; i++) {
        const txnBuffer = Buffer.from(data[i], "base64");
        const txn = VersionedTransaction.deserialize(txnBuffer);
        txn.sign([wallets[i]]);
        signatures.push(bs58.encode(txn.signatures[0]));
        const signedTxnBuffer = txn.serialize();
        encodedSignedTxns.push(bs58.encode(signedTxnBuffer));
      }

      try {
        const jitoResponse = await fetch(
          `https://tokyo.mainnet.block-engine.jito.wtf/api/v1/bundles`,
          {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify({
              jsonrpc: "2.0",
              id: 1,
              method: "sendBundle",
              params: [encodedSignedTxns],
            }),
          }
        );
        if (jitoResponse.status === 200) {
          for (let i = 0; i < signatures.length; i++) {
            console.log(
              `Transaction ${i}: https://solscan.io/tx/${signatures[i]}`
            );
          }
        } else {
          console.log(
            "bundle failed, please check the parameters",
            jitoResponse.statusText
          );
        }
      } catch (e: any) {
        console.error(e.message);
      }
    } else {
      console.log(response.statusText); // log error
    }
  } catch (e: any) {
    console.error(e.message);
  }
};

test();

File: python/createToken/create.buy.bundle.token.pumpfun.test.py
#!/usr/bin/env python3
from dotenv import load_dotenv
import os
import base64
import base58
import requests
from solana.keypair import Keypair
from solana.transaction import VersionedTransaction

def main():
    # Load your PRIVATE_KEY from .env
    load_dotenv()
    private_key_b58 = os.getenv("PRIVATE_KEY")
    if not private_key_b58:
        raise EnvironmentError("Missing PRIVATE_KEY in environment")

    # Decode secret and build 5 identical Keypairs
    secret = base58.b58decode(private_key_b58)
    wallets = [Keypair.from_secret_key(secret) for _ in range(5)]

    # 1) Upload metadata + image to IPFS (pump.fun)
    ipfs_url = "https://pump.fun/api/ipfs"
    files = {"file": open("./zz.png", "rb")}
    data = {
        "name": "test",
        "symbol": "TEST",
        "description": "This is an example token created via SolanaPortal.AI",
        "twitter": "https://x.com/solana",
        "telegram": "https://x.com/solana",
        "website": "https://solanaportal.ai",
        "showName": "true",
    }
    ipfs_resp = requests.post(ipfs_url, files=files, data=data)
    if ipfs_resp.status_code != 200:
        print("Failed to create IPFS metadata storage:", ipfs_resp.text)
        return

    ipfs_json = ipfs_resp.json()
    metadata     = ipfs_json["metadata"]
    metadata_uri = ipfs_json["metadataUri"]

    # 2) Build the bundle-creation payload
    create_url = "https://api.solanaportal.io/api/create/token/pumpfun"
    params = [
        {
            "name":         metadata["name"],
            "symbol":       metadata["symbol"],
            "metadataUri":  metadata_uri,
            "wallet_address": str(wallets[0].public_key),
            "amount":       0.00001,
            "slippage":     100,
            "tip":          0.0001,
            "type":         "jito",
        },
        {
            "wallet_address": str(wallets[1].public_key),
            "action":         "buy",
            "amount":         0.000001,
        },
        {
            "wallet_address": str(wallets[2].public_key),
            "action":         "buy",
            "amount":         0.000002,
        },
        {
            "wallet_address": str(wallets[3].public_key),
            "action":         "buy",
            "amount":         0.000003,
        },
        {
            "wallet_address": str(wallets[4].public_key),
            "action":         "buy",
            "amount":         0.000004,
        },
    ]
    create_resp = requests.post(create_url, json=params)
    if create_resp.status_code != 200:
        print("Error fetching bundle TXNs:", create_resp.text)
        return

    # 3) Deserialize, sign each TXN, and collect signatures
    unsigned_b64 = create_resp.json()  # list of base64-encoded TXNs
    signed_b58   = []
    signatures   = []
    for i, txn_b64 in enumerate(unsigned_b64):
        txn_bytes = base64.b64decode(txn_b64)
        txn       = VersionedTransaction.deserialize(txn_bytes)
        txn.sign([wallets[i]])
        # record first signature for logging
        sig = base58.b58encode(txn.signatures[0]).decode()
        signatures.append(sig)
        # serialize and base58-encode
        signed_bytes = txn.serialize()
        signed_b58.append(base58.b58encode(signed_bytes).decode())

    # 4) Send the signed bundle to Jito
    jito_url = "https://tokyo.mainnet.block-engine.jito.wtf/api/v1/bundles"
    payload = {
        "jsonrpc": "2.0",
        "id":      1,
        "method":  "sendBundle",
        "params":  [signed_b58],
    }
    jito_resp = requests.post(jito_url, json=payload)
    if jito_resp.status_code == 200:
        for idx, sig in enumerate(signatures):
            print(f"Transaction {idx}: https://solscan.io/tx/{sig}")
    else:
        print("Bundle failed:", jito_resp.text)

if __name__ == "__main__":
    main()
